src/common/common_service/src/observer_manager.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::Duration;
+
+use risingwave_common::error::{ErrorCode, Result};
+use risingwave_common::util::addr::HostAddr;
+use risingwave_pb::common::WorkerType;
+use risingwave_pb::meta::SubscribeResponse;
+use risingwave_rpc_client::{MetaClient, NotificationStream};
+use tokio::task::JoinHandle;
+
+/// `ObserverManager` is used to update data based on notification from meta.
+/// Call `start` to spawn a new asynchronous task
+/// We can write the notification logic by implementing `ObserverNodeImpl`.
+pub struct ObserverManager {
+    rx: Box<dyn NotificationStream>,
+    meta_client: MetaClient,
+    addr: HostAddr,
+    worker_type: WorkerType,
+    observer_states: Box<dyn ObserverNodeImpl + Send>,
+}
+
+pub trait ObserverNodeImpl {
+    /// modify data after receiving notification from meta
+    fn handle_notification(&mut self, resp: SubscribeResponse);
+
+    /// Initialize data from the meta. It will be called at start or resubscribe
+    fn handle_snapshot_notification(&mut self, resp: SubscribeResponse) -> Result<()>;
```

Comment:
A small naming issue. The `snapshot` in `handle_snapshot_notification` may be confused with hummock snapshot concept?

Reply:
I will change it.